### PR TITLE
feat(client): allow accelerate to hook its own engine

### PR DESCRIPTION
**Pull request by millsp**
This PR enables Accelerate to manage its own implementation of an engine, this means we can transfer the ownership of it. This contributes to better gzip sizes. This PR will be followed by another one in order to fully remove `DataProxyEngine`.

/integration

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -1,3 +1,7 @@
+const path = require('path')
+
+const project = path.resolve(process.cwd(), 'tsconfig.json')
```
Comment (millsp) [inline .eslintrc.js:3]: Unrelated: Fix for very slow eslint autosave in vscode.

File: packages/client/src/runtime/core/extensions/$extends.ts [hunk 2]
```diff
@@ -13,10 +11,12 @@ export function $extends(this: Client, extension: ExtensionArgs | ((client: Clie
     return extension(this)
   }
 
-  // re-apply models to the extend client: they always capture specific instance
-  // of the client and without re-application they would not see new extensions
-  const oldClient = unApplyModelsAndClientExtensions(this)
-  const newClient = Object.create(oldClient, {
+  if (extension.client?.__AccelerateEngine) {
+    const Engine = extension.client.__AccelerateEngine as typeof AccelerateEngine
+    this._originalClient._engine = new Engine(this._originalClient._accelerateEngineConfig) as any
```
Comment (millsp) [inline packages/client/src/runtime/core/extensions/$extends.ts:16]:
```suggestion
    this._originalClient._engine = new Engine(this._originalClient._accelerateEngineConfig)
```

File: packages/accelerate-contract/package.json [hunk 3]
```diff
@@ -0,0 +1,27 @@
+{
+  "name": "@prisma/accelerate-contract",
+  "version": "0.0.0",
+  "description": "This package is intended for Prisma's internal use",
+  "main": "dist/index.js",
+  "types": "dist/index.d.ts",
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/prisma/prisma.git",
+    "directory": "packages/engines"
```
Comment (Jolg42) [inline packages/accelerate-contract/package.json:10]:
```suggestion
    "directory": "packages/accelerate-contract"
```

File: packages/client/src/runtime/core/engines/accelerate/AccelerateEngine.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
+/* eslint-disable @typescript-eslint/no-namespace */
+/* eslint-disable @typescript-eslint/no-non-null-asserted-optional-chain */
+import { PrismaClientInitializationError } from '../../errors/PrismaClientInitializationError'
+import { BatchQueryEngineResult, Engine, EngineConfig, RequestBatchOptions, RequestOptions } from '../common/Engine'
+import { JsonQuery } from '../common/types/JsonProtocol'
+import { Metrics, MetricsOptionsJson, MetricsOptionsPrometheus } from '../common/types/Metrics'
+import { QueryEngineResult } from '../common/types/QueryEngine'
+import { InteractiveTransactionInfo as ITXInfo, Options, TransactionHeaders } from '../common/types/Transaction'
+
+const ERROR_MESSAGE = `Accelerate has not been setup correctly. Make sure your client is using \`.$extends(withAccelerate())\`.`
```
Comment (Jolg42) [inline packages/client/src/runtime/core/engines/accelerate/AccelerateEngine.ts:10]:
```suggestion
const ERROR_MESSAGE = `Accelerate has not been setup correctly. Make sure your client is using \`.$extends(withAccelerate())\`. See https://pris.ly/d/accelerate-getting-started`
```

Comment (Jolg42) [reply packages/client/src/runtime/core/engines/accelerate/AccelerateEngine.ts:10]: Found in dub

**Review by Jolg42 — APPROVED** [review]

**Pull request merged**